chore: migrate to TypeScript

gclapperton opened the pull request:
#### What does this PR do?

Migrates the library to TypeScript. Note: this PR is about _enabling_ TypeScript and not leveraging it. Where the compiler complained about types, I just added `any`... for now. And it does also not yet generate type declarations. This was done intentionally as I didn't want to break consumer if they started relying on them while they were still being in flux

- imports/exports have been switched to ESM
- `.npmignore` has been deleted in favour of explicitly specifying files to be included using the `files` key in the `package.json`
- formatted with `nom run format` (prettier)

#### How should this be manually tested?

`npm run test` to make sure that the library is not broken
`npm pack` to see what will be in the published package

`npm link` and then `npm link snyk-policy` in Registry before running Registry tests to make sure the clients can still consume the library and everything is working as expected

gclapperton commented on package.json:
```diff
@@ -43,7 +56,7 @@
   "tap": {
     "branches": 85,
-    "functions": 95,
+    "functions": 90,
```
Coverage does not appear to be stable. Modifying imports seems to affect coverage, but it's not clear why. We didn't have thresholds before, so dropping to 90% of functions

cat2608 commented on lib/add-exclude.ts:
```diff
@@ -1,6 +1,6 @@
-module.exports = addExclude;
+export default addExclude;
```
Not blocking: `export default` is a matter of personal preference and coding style but it can make it unclear which parts of a module are being exported, particularly when multiple defaults are exported from the same file.

Default exports, allow you to export a single functions from a module without using a specific name and then, when importing a default export, any name can be chosen. Sometimes it's a bit confusing when it happens.

Of course, if we chose to use named exports, we'll need to update the imports, for example from `import addExclude from './add-exclude';` to `import { addExclude } from './add-exclude';` -- out of scope of this PR I believe

cat2608 commented on lib/index.ts:
```diff
@@ -117,7 +110,7 @@ function load(root, options) {
           if (debug.enabled) {
             debug('final policy:');
-            debug(JSON.stringify(res, '', 2));
+            debug(JSON.stringify(res, [], 2));
```
The signature of `JSON.stringify` is like `JSON.stringify(value, replacer, space)`.
I think we need t replace the `''` with null since if we leave the `''` the output is converted to:

```node
const person = {
  firstName: 'John',
  age: 30,
  address: {
    state: 'CA',
    zip: '12345'
  }
}; 
JSON.stringify(person, [], 2)
"{}" 

// vs

JSON.stringify(person, null, 2)
'{
  "firstName": "John",
  "age": 30,
  "address": {
    "state": "CA",
    "zip": "12345"
  }
}' 

```

gclapperton replied:
Good catch! Just tested and `JSON.stringify(res, null, 2)` does give the same result as `JSON.stringify(res, '', 2)` 🙏

cat2608 replied:
I forgot to point out that: 

>Replacer can be either a function or an array.
>A function that alters the behaviour of the stringification process, or an array of strings and numbers that specifies properties of value to be included in the output.

[source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)

gclapperton replied:
Is `''` a function or an array?!!! :p

cat2608 reviewed (approved):
Thank you 🙌

gclapperton commented on lib/match.ts:
```diff
@@ -166,7 +165,7 @@ function matchExactWithStars(vuln, path) {
 }
 
-function getByVuln(policy, vuln) {
+function getByVuln(policy?, vuln?) {
```
It gets exported as part of the libraries public API [here](https://github.com/snyk/policy/pull/101/files#diff-d5a4b3a0309f2337144861084c1014523cced561eb722cb9684c63f5e41e0567R11)

gclapperton committed:
chore: migrate to TypeScript

BREAKING CHANGE: Typescript migration is intended to be a drop in replacement, but marked as a breaking change to be safe

gclapperton commented on package.json:
```diff
@@ -1,15 +1,25 @@
-    "check-tests": "! grep 'test.only' test/**/*.test.js -n",
-    "env": "node -e 'console.log(process.env, process.versions)'",
-    "tap": "tap test/**/*.test.js -R spec --timeout=60",
+    "prepack": "npm run build",
```
`prepublish` is deprecated (and doesn't do what I thought it did). I've updated to use `prepack` instead https://docs.npmjs.com/cli/v7/using-npm/scripts#life-cycle-scripts

dekelund reviewed (approved):
Reviewed lib/add-exclude.js and it looks sane

snyksec commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/snyk-policy/v/2.0.0)
- [GitHub release](https://github.com/snyk/policy/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
